src/main/java/com/tabnine/binary/fetch/BootstrapperSupport.java

```diff
@@ -0,0 +1,73 @@
+package com.tabnine.binary.fetch;
+
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.util.text.SemVer;
+import com.tabnine.general.StaticConfig;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.prefs.Preferences;
+
+import static java.lang.String.format;
+
+public class BootstrapperSupport {
+    static Optional<BinaryVersion> bootstrap_version(LocalBinaryVersions localBinaryVersions, BinaryRemoteSource binaryRemoteSource, BinaryDownloader binaryDownloader) {
+        // bootstrapper can change versions under the hood
+        // plugin must ensure it is running a version containing the bootstrapper but not override any versions that the bootstrapper may have downloaded on its own
+        // we persist the first known bootstrapper version and choose accordingly
+
+        Optional<SemVer> thisVersion = minimalBootstrappedVersion();
+
+        if (thisVersion.isPresent()) {
```

Comment:
prefer map & orElseGet, this is a general comment for when using optional

Reply:
obviously. there is additional logic here and I don't like the functional style that would be nested